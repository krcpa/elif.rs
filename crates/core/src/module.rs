use crate::container::{Container, ContainerBuilder};
use crate::provider::{ProviderRegistry, ServiceProvider};
use thiserror::Error;

/// Application module trait that integrates with service-builder
pub trait Module: Send + Sync {
    /// Module name for identification
    fn name(&self) -> &'static str;
    
    /// Configure services in the container builder
    fn configure(&self, builder: ContainerBuilder) -> Result<ContainerBuilder, ModuleError>;
    
    /// Boot the module after container is built
    fn boot(&self, _container: &Container) -> Result<(), ModuleError> {
        // Default implementation does nothing
        Ok(())
    }
    
    /// Module dependencies (other modules that must be loaded first)
    fn dependencies(&self) -> Vec<&'static str> {
        vec![]
    }
}

/// Application that manages modules, providers, and container
pub struct Application {
    container: Container,
    modules: Vec<Box<dyn Module>>,
    providers: ProviderRegistry,
}

impl Application {
    /// Create a new application builder
    pub fn builder() -> ApplicationBuilder {
        ApplicationBuilder::new()
    }
    
    /// Get the service container
    pub fn container(&self) -> &Container {
        &self.container
    }
    
    /// Start the application by booting providers and modules
    pub async fn start(&mut self) -> Result<(), ApplicationError> {
        // Boot all providers first
        self.providers.boot_all(&self.container)
            .map_err(ApplicationError::ProviderBoot)?;
        
        // Then boot all modules
        for module in &self.modules {
            module.boot(&self.container)
                .map_err(|e| ApplicationError::ModuleBoot {
                    module: module.name().to_string(),
                    error: e,
                })?;
        }
        Ok(())
    }
}

/// Builder for constructing applications
pub struct ApplicationBuilder {
    modules: Vec<Box<dyn Module>>,
    providers: ProviderRegistry,
}

impl ApplicationBuilder {
    pub fn new() -> Self {
        Self {
            modules: vec![],
            providers: ProviderRegistry::new(),
        }
    }
    
    /// Add a module to the application
    pub fn module<M: Module + 'static>(mut self, module: M) -> Self {
        self.modules.push(Box::new(module));
        self
    }
    
    /// Add a service provider to the application
    pub fn provider<P: ServiceProvider + 'static>(mut self, provider: P) -> Self {
        self.providers.register(provider);
        self
    }
    
    /// Build the application by resolving dependencies and creating container
    pub fn build(mut self) -> Result<Application, ApplicationError> {
        // Resolve provider dependencies first
        self.providers.resolve_dependencies()
            .map_err(ApplicationError::ProviderDependency)?;
        
        // Create container with providers first
        let mut builder = Container::builder();
        
        // Register all providers
        builder = self.providers.register_all(builder)
            .map_err(ApplicationError::ProviderRegistration)?;
        
        // Sort modules by dependencies and configure them
        let sorted_modules = self.resolve_module_dependencies(builder)?;
        
        Ok(sorted_modules)
    }
    
    /// Resolve module dependencies and build container
    fn resolve_module_dependencies(self, mut builder: ContainerBuilder) -> Result<Application, ApplicationError> {
        let modules = self.modules;
        
        // For now, just configure modules in order (full dependency resolution can be added later)
        for module in &modules {
            builder = module.configure(builder)
                .map_err(|e| ApplicationError::ModuleConfiguration {
                    module: module.name().to_string(),
                    error: e,
                })?;
        }
        
        let container = builder.build()
            .map_err(|e| ApplicationError::ContainerBuild {
                error: e.to_string(),
            })?;
        
        Ok(Application {
            container,
            modules,
            providers: self.providers,
        })
    }
    
}

#[derive(Error, Debug)]
pub enum ModuleError {
    #[error("Module configuration failed: {message}")]
    ConfigurationFailed { message: String },
    
    #[error("Module boot failed: {message}")]
    BootFailed { message: String },
    
    #[error("Service registration failed: {service}")]
    ServiceRegistrationFailed { service: String },
}

#[derive(Error, Debug)]
pub enum ApplicationError {
    #[error("Module configuration error in '{module}': {error}")]
    ModuleConfiguration { module: String, error: ModuleError },
    
    #[error("Module boot error in '{module}': {error}")]
    ModuleBoot { module: String, error: ModuleError },
    
    #[error("Container build failed: {error}")]
    ContainerBuild { error: String },
    
    #[error("Circular dependency detected for module '{module}'")]
    CircularDependency { module: String },
    
    #[error("Missing dependency '{dependency}' for module '{module}'")]
    MissingDependency { module: String, dependency: String },
    
    #[error("Provider dependency error: {0}")]
    ProviderDependency(#[from] crate::provider::ProviderError),
    
    #[error("Provider registration error: {0}")]
    ProviderRegistration(crate::provider::ProviderError),
    
    #[error("Provider boot error: {0}")]
    ProviderBoot(crate::provider::ProviderError),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::container::{AppConfig, DatabaseConnection, Environment};
    use std::sync::Arc;
    
    // Test implementations
    struct TestConfig;
    impl AppConfig for TestConfig {
        fn get(&self, _key: &str) -> Option<String> { None }
        fn environment(&self) -> Environment { Environment::Testing }
    }
    
    struct TestDatabase;
    impl DatabaseConnection for TestDatabase {
        fn is_connected(&self) -> bool { true }
        fn execute(&self, _query: &str) -> Result<(), crate::container::DatabaseError> { Ok(()) }
    }
    
    // Test modules
    struct CoreModule;
    impl Module for CoreModule {
        fn name(&self) -> &'static str { "core" }
        
        fn configure(&self, builder: ContainerBuilder) -> Result<ContainerBuilder, ModuleError> {
            let config = Arc::new(TestConfig) as Arc<dyn AppConfig>;
            let database = Arc::new(TestDatabase) as Arc<dyn DatabaseConnection>;
            
            Ok(builder.config(config).database(database))
        }
    }
    
    struct AuthModule;
    impl Module for AuthModule {
        fn name(&self) -> &'static str { "auth" }
        
        fn dependencies(&self) -> Vec<&'static str> {
            vec!["core"]
        }
        
        fn configure(&self, builder: ContainerBuilder) -> Result<ContainerBuilder, ModuleError> {
            // Auth module doesn't add new services, just uses existing ones
            Ok(builder)
        }
        
        fn boot(&self, container: &Container) -> Result<(), ModuleError> {
            let _config = container.config();
            // Perform auth module initialization
            Ok(())
        }
    }
    
    #[tokio::test]
    async fn test_application_with_modules_and_providers() {
        use crate::provider::ServiceProvider;
        
        // Create a provider that handles service registration
        struct TestProvider;
        impl ServiceProvider for TestProvider {
            fn name(&self) -> &'static str { "test" }
            
            fn register(&self, builder: crate::container::ContainerBuilder) -> Result<crate::container::ContainerBuilder, crate::provider::ProviderError> {
                let config = Arc::new(TestConfig) as Arc<dyn AppConfig>;
                let database = Arc::new(TestDatabase) as Arc<dyn DatabaseConnection>;
                Ok(builder.config(config).database(database))
            }
        }
        
        let mut app = Application::builder()
            .provider(TestProvider)
            .module(AuthModule)
            .build()
            .unwrap();
        
        // Verify container is properly configured
        let config = app.container().config();
        assert_eq!(config.environment() as u8, Environment::Testing as u8);
        
        // Start the application
        app.start().await.unwrap();
    }
    
    #[test]
    fn test_provider_integration() {
        use crate::provider::ServiceProvider;
        
        struct TestProvider;
        impl ServiceProvider for TestProvider {
            fn name(&self) -> &'static str { "test" }
            
            fn register(&self, builder: crate::container::ContainerBuilder) -> Result<crate::container::ContainerBuilder, crate::provider::ProviderError> {
                let config = Arc::new(TestConfig) as Arc<dyn AppConfig>;
                let database = Arc::new(TestDatabase) as Arc<dyn DatabaseConnection>;
                Ok(builder.config(config).database(database))
            }
        }
        
        let result = Application::builder()
            .provider(TestProvider)
            .build();
        
        assert!(result.is_ok());
    }
}