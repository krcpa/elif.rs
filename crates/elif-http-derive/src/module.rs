//! Module system macro implementation
//! 
//! Provides the `#[module(...)]` attribute macro and `module! { ... }` function-like macro
//! for defining dependency injection modules and application composition.
//!
//! Features:
//! - Provider definitions with trait mappings: `EmailService => SmtpEmailService @ "smtp"`
//! - Controller registration: `controllers: [Controller1, Controller2]`
//! - Module imports/exports: `imports: [Module], exports: [Service]`
//! - Application composition: `app! { modules: [Module1, Module2] }`

use proc_macro::TokenStream;
use proc_macro2::{Ident, Span};
use quote::quote;
use syn::{
    parse::{Parse, ParseStream, Result},
    parse_macro_input,
    punctuated::Punctuated,
    token::{Comma, FatArrow, At},
    Error, ItemStruct, Type, LitStr, Token,
};

/// Main implementation function for the module attribute macro
pub fn module_impl(args: TokenStream, input: TokenStream) -> TokenStream {
    let module_args = match syn::parse::<ModuleArgs>(args) {
        Ok(args) => args,
        Err(err) => return err.to_compile_error().into(),
    };
    
    let mut item_struct = parse_macro_input!(input as ItemStruct);
    
    match process_module_attribute(&mut item_struct, module_args) {
        Ok(result) => result.into(),
        Err(err) => err.to_compile_error().into(),
    }
}

/// Main implementation function for the module function-like macro
pub fn module_composition_impl(input: TokenStream) -> TokenStream {
    let composition_args = match syn::parse::<ModuleCompositionArgs>(input) {
        Ok(args) => args,
        Err(err) => return err.to_compile_error().into(),
    };
    
    match generate_application_composition(composition_args) {
        Ok(result) => result.into(),
        Err(err) => err.to_compile_error().into(),
    }
}

/// Arguments parsed from the #[module(...)] attribute
#[derive(Debug, Clone)]
pub struct ModuleArgs {
    pub providers: Vec<ProviderDef>,
    pub controllers: Vec<Type>,
    pub imports: Vec<Type>,
    pub exports: Vec<Type>,
}

impl Parse for ModuleArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut providers = Vec::new();
        let mut controllers = Vec::new();
        let mut imports = Vec::new();
        let mut exports = Vec::new();
        
        // Parse comma-separated key-value pairs
        while !input.is_empty() {
            let key: Ident = input.parse()?;
            let _colon: Token![:] = input.parse()?;
            
            let key_str = key.to_string();
            match key_str.as_str() {
                "providers" => {
                    providers = parse_provider_list(input)?;
                },
                "controllers" => {
                    controllers = parse_type_list(input)?;
                },
                "imports" => {
                    imports = parse_type_list(input)?;
                },
                "exports" => {
                    exports = parse_type_list(input)?;
                },
                _ => {
                    return Err(Error::new_spanned(
                        key,
                        format!(
                            "Unknown module section '{}'. Valid sections are: providers, controllers, imports, exports",
                            key_str
                        )
                    ));
                }
            }
            
            // Optional comma between sections
            if !input.is_empty() {
                let _comma: Option<Comma> = input.parse().ok();
            }
        }
        
        Ok(ModuleArgs {
            providers,
            controllers,
            imports,
            exports,
        })
    }
}

/// Definition of a provider in the module
/// Supports various patterns:
/// - `UserService` (concrete service)
/// - `EmailService => SmtpEmailService` (trait mapping)
/// - `EmailService => SmtpEmailService @ "smtp"` (named trait mapping)
/// - `dyn EmailService => SmtpEmailService` (explicit dyn syntax still supported)
#[derive(Debug, Clone)]
pub struct ProviderDef {
    pub service_type: ProviderType,
    pub implementation: Option<Type>,
    pub name: Option<String>,
}

#[derive(Debug, Clone)]
pub enum ProviderType {
    /// Concrete service type: UserService
    Concrete(Type),
    /// Trait type: dyn EmailService
    Trait(Type),
}

impl Parse for ProviderDef {
    fn parse(input: ParseStream) -> Result<Self> {
        // Parse the service type (may be dyn Trait, bare Trait, or concrete type)
        let service_type = if input.peek(Token![dyn]) {
            // Explicit dyn Trait syntax
            let _dyn: Token![dyn] = input.parse()?;
            let trait_type: Type = input.parse()?;
            ProviderType::Trait(trait_type)
        } else {
            let parsed_type: Type = input.parse()?;
            
            // Check if this will be followed by => (trait mapping)
            if input.peek(FatArrow) {
                // If there's a =>, it's a trait mapping, so treat as trait
                ProviderType::Trait(parsed_type)
            } else {
                // No =>, so it's a concrete service
                ProviderType::Concrete(parsed_type)
            }
        };
        
        let mut implementation = None;
        let mut name = None;
        
        // Check for trait mapping: => Implementation
        if input.peek(FatArrow) {
            let _arrow: FatArrow = input.parse()?;
            implementation = Some(input.parse()?);
            
            // Check for named mapping: @ "name"
            if input.peek(At) {
                let _at: At = input.parse()?;
                let name_lit: LitStr = input.parse()?;
                name = Some(name_lit.value());
            }
        }
        
        Ok(ProviderDef {
            service_type,
            implementation,
            name,
        })
    }
}

/// Parse a list of providers: [Provider1, dyn Trait => Impl, ...]
fn parse_provider_list(input: ParseStream) -> Result<Vec<ProviderDef>> {
    let content;
    let _bracket = syn::bracketed!(content in input);
    let providers: Punctuated<ProviderDef, Comma> = content.parse_terminated(ProviderDef::parse, Comma)?;
    Ok(providers.into_iter().collect())
}

/// Parse a list of types: [Type1, Type2, ...]
fn parse_type_list(input: ParseStream) -> Result<Vec<Type>> {
    let content;
    let _bracket = syn::bracketed!(content in input);
    let types: Punctuated<Type, Comma> = content.parse_terminated(Type::parse, Comma)?;
    Ok(types.into_iter().collect())
}

/// Arguments for module composition macro: module! { ... }
#[derive(Debug, Clone)]
pub struct ModuleCompositionArgs {
    pub modules: Vec<Type>,
    pub overrides: Vec<ProviderDef>,
}

impl Parse for ModuleCompositionArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut modules = Vec::new();
        let mut overrides = Vec::new();
        
        while !input.is_empty() {
            let key: Ident = input.parse()?;
            let _colon: Token![:] = input.parse()?;
            
            let key_str = key.to_string();
            match key_str.as_str() {
                "modules" => {
                    modules = parse_type_list(input)?;
                },
                "overrides" => {
                    overrides = parse_provider_list(input)?;
                },
                _ => {
                    return Err(Error::new_spanned(
                        key,
                        format!(
                            "Unknown composition section '{}'. Valid sections are: modules, overrides",
                            key_str
                        )
                    ));
                }
            }
            
            if !input.is_empty() {
                let _comma: Option<Comma> = input.parse().ok();
            }
        }
        
        if modules.is_empty() {
            return Err(Error::new(
                Span::call_site(),
                "module! composition requires at least one module in the 'modules' section"
            ));
        }
        
        Ok(ModuleCompositionArgs {
            modules,
            overrides,
        })
    }
}

/// Process the module attribute and generate module registration code
fn process_module_attribute(
    item_struct: &mut ItemStruct,
    module_args: ModuleArgs,
) -> Result<proc_macro2::TokenStream> {
    let struct_name = &item_struct.ident;
    
    // Generate module descriptor method
    let module_descriptor_impl = generate_module_descriptor_method(struct_name, &module_args)?;
    
    Ok(quote! {
        #item_struct
        
        #module_descriptor_impl
    })
}

/// Generate module descriptor method for runtime registration
fn generate_module_descriptor_method(
    struct_name: &Ident,
    module_args: &ModuleArgs,
) -> Result<proc_macro2::TokenStream> {
    let providers_code = generate_providers_registration(&module_args.providers)?;
    let controllers_code = generate_controllers_registration(&module_args.controllers)?;
    let imports_code = generate_imports_registration(&module_args.imports)?;
    let exports_code = generate_exports_registration(&module_args.exports)?;
    
    let descriptor_name = quote::format_ident!("{}ModuleDescriptor", struct_name);
    
    Ok(quote! {
        impl #struct_name {
            /// Get the module descriptor for this module
            /// Note: This is a stub implementation for Epic 1 (Parser Foundation)
            /// Full integration with elif-core will be implemented in Epic 4 (Runtime Integration)
            pub fn module_descriptor() -> #descriptor_name {
                let mut descriptor = #descriptor_name::new(stringify!(#struct_name));
                
                #providers_code
                #controllers_code
                #imports_code
                #exports_code
                
                descriptor
            }
        }
        
        /// Stub implementation for module descriptor - will be replaced with real elif-core types in Epic 4
        #[derive(Debug, Clone)]
        pub struct #descriptor_name {
            name: String,
        }
        
        impl #descriptor_name {
            pub fn new(name: &str) -> Self {
                Self {
                    name: name.to_string(),
                }
            }
            
            pub fn name(&self) -> &str {
                &self.name
            }
        }
    })
}

/// Generate provider registration code  
fn generate_providers_registration(providers: &[ProviderDef]) -> Result<proc_macro2::TokenStream> {
    if providers.is_empty() {
        return Ok(quote! {
            // No providers specified
        });
    }
    
    let mut comments = Vec::new();
    
    for provider in providers {
        let comment = match &provider.service_type {
            ProviderType::Concrete(_service_type) => {
                match &provider.name {
                    Some(_name) => {
                        quote! {
                            // Provider: #service_type (named: #name)
                        }
                    },
                    None => {
                        quote! {
                            // Provider: #service_type
                        }
                    }
                }
            },
            ProviderType::Trait(trait_type) => {
                match &provider.implementation {
                    Some(_impl_type) => {
                        match &provider.name {
                            Some(_name) => {
                                quote! {
                                    // Provider: dyn #trait_type => #impl_type (named: #name)
                                }
                            },
                            None => {
                                quote! {
                                    // Provider: dyn #trait_type => #impl_type
                                }
                            }
                        }
                    },
                    None => {
                        return Err(Error::new_spanned(
                            trait_type,
                            "Trait providers must specify implementation type: dyn Trait => Implementation"
                        ));
                    }
                }
            }
        };
        
        comments.push(comment);
    }
    
    Ok(quote! {
        #(#comments)*
    })
}

/// Generate controller registration code
fn generate_controllers_registration(controllers: &[Type]) -> Result<proc_macro2::TokenStream> {
    if controllers.is_empty() {
        return Ok(quote! {
            // No controllers specified
        });
    }
    
    let comments: Vec<_> = controllers.iter().map(|_controller| {
        quote! {
            // Controller: #controller
        }
    }).collect();
    
    Ok(quote! {
        #(#comments)*
    })
}

/// Generate imports registration code  
fn generate_imports_registration(imports: &[Type]) -> Result<proc_macro2::TokenStream> {
    if imports.is_empty() {
        return Ok(quote! {
            // No imports specified
        });
    }
    
    let comments: Vec<_> = imports.iter().map(|_import| {
        quote! {
            // Import: #import
        }
    }).collect();
    
    Ok(quote! {
        #(#comments)*
    })
}

/// Generate exports registration code
fn generate_exports_registration(exports: &[Type]) -> Result<proc_macro2::TokenStream> {
    if exports.is_empty() {
        return Ok(quote! {
            // No exports specified
        });
    }
    
    let comments: Vec<_> = exports.iter().map(|_export| {
        quote! {
            // Export: #export
        }
    }).collect();
    
    Ok(quote! {
        #(#comments)*
    })
}

/// Generate application composition code
fn generate_application_composition(
    composition_args: ModuleCompositionArgs,
) -> Result<proc_macro2::TokenStream> {
    let modules_registration = generate_modules_registration(&composition_args.modules)?;
    let overrides_registration = generate_overrides_registration(&composition_args.overrides)?;
    
    Ok(quote! {
        {
            // Stub implementation for Epic 1 (Parser Foundation)
            // Full runtime integration will be implemented in Epic 4
            
            #modules_registration
            #overrides_registration
            
            // Return placeholder
            ()
        }
    })
}

/// Generate modules registration for application composition
fn generate_modules_registration(modules: &[Type]) -> Result<proc_macro2::TokenStream> {
    let comments: Vec<_> = modules.iter().map(|_module| {
        quote! {
            // Module: #module
        }
    }).collect();
    
    Ok(quote! {
        #(#comments)*
    })
}

/// Generate overrides registration for application composition
fn generate_overrides_registration(overrides: &[ProviderDef]) -> Result<proc_macro2::TokenStream> {
    if overrides.is_empty() {
        return Ok(quote! {
            // No overrides specified
        });
    }
    
    let mut comments = Vec::new();
    
    for override_def in overrides {
        let comment = match &override_def.service_type {
            ProviderType::Concrete(_service_type) => {
                match &override_def.name {
                    Some(_name) => {
                        quote! {
                            // Override: #service_type (named: #name)
                        }
                    },
                    None => {
                        quote! {
                            // Override: #service_type
                        }
                    }
                }
            },
            ProviderType::Trait(trait_type) => {
                match &override_def.implementation {
                    Some(_impl_type) => {
                        match &override_def.name {
                            Some(_name) => {
                                quote! {
                                    // Override: dyn #trait_type => #impl_type (named: #name)
                                }
                            },
                            None => {
                                quote! {
                                    // Override: dyn #trait_type => #impl_type
                                }
                            }
                        }
                    },
                    None => {
                        return Err(Error::new_spanned(
                            trait_type,
                            "Trait overrides must specify implementation type: dyn Trait => Implementation"
                        ));
                    }
                }
            }
        };
        
        comments.push(comment);
    }
    
    Ok(quote! {
        #(#comments)*
    })
}