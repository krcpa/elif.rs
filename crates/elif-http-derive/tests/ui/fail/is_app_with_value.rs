//! Test that is_app: true causes a compilation error
//! is_app should only be used as a bare flag without a colon or value

use elif_http_derive::module;

#[derive(Default)]
pub struct TestService;

#[derive(Default)]
pub struct TestController;

// This should fail - is_app should be used without a value
#[module(
    providers: [TestService],
    controllers: [TestController],
    is_app: true
)]
pub struct BadAppModule;

fn main() {}