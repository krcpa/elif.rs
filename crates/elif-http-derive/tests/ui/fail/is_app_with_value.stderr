error: Unknown module section 'is_app'. Valid sections are: providers, controllers, imports, exports, is_app.

       ğŸ’¡ Suggestions:
       â€¢ Use 'providers: [ServiceType]' for concrete services
       â€¢ Use 'providers: [dyn Trait => Implementation]' for trait mappings
       â€¢ Use 'controllers: [ControllerType]' for HTTP controllers
       â€¢ Use 'imports: [ModuleType]' for module dependencies
       â€¢ Use 'exports: [ServiceType]' for services available to other modules
       â€¢ Use 'is_app' (without colon) to mark this module as an app module that can bootstrap

       ğŸ“– See: https://docs.elif.rs/modules/module-definition
  --> tests/ui/fail/is_app_with_value.rs:16:5
   |
16 |     is_app: true
   |     ^^^^^^
